internal/flowrate/util.go

```diff
@@ -7,24 +7,50 @@ package flowrate
 import (
 	"math"
 	"strconv"
+	"sync/atomic"
 	"time"
 )
 
 // clockRate is the resolution and precision of clock().
 const clockRate = 20 * time.Millisecond
 
-// czero is the process start time rounded down to the nearest clockRate
-// increment.
-var czero = time.Now().Round(clockRate)
+var (
+	hasInitializedClock = atomic.Bool{}
+	currentClockValue   = atomic.Int64{}
+	clockStartTime      = time.Time{}
+)
+
+// checks if the clock update timer is running. If not, sets clockStartTime and starts it.
+func ensureClockRunning() {
+	firstRun := hasInitializedClock.CompareAndSwap(false, true)
+	if !firstRun {
+		return
+	}
+	clockStartTime = time.Now().Round(clockRate)
+	go runClockUpdates()
+}
+
+// increments the current clock value every clockRate interval.
+func runClockUpdates() {
+	// sleep, then increment the clock value
+	// This is the only place the clock value is updated.
+	// Ensures that the clock value starts at 0.
+	for {
```

Comment:
not sure how I feel about us not ever exiting here. I guess it's fine.

Reply:
I can use some more CAS instructions to make it exit

Reply:
EDIT didn't work